src/node_zip.cc

```diff
@@ -0,0 +1,454 @@
+  static void DeleteEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void ReadEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void StatEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void RestatEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void Digest(const v8::FunctionCallbackInfo<v8::Value>& args);
+
+  int GetFileType(zip_uint64_t index) const;
+
+  std::vector<char> buf_;
+  zip_error_t error_;
+  zip_source_t* source_ = nullptr;
+  zip_t* zip_ = nullptr;
+
+  SET_SELF_SIZE(ZipArchive)
+};
+
+ZipArchive::ZipArchive(Environment* env, v8::Local<v8::Object> obj, std::vector<char>&& buf)
+    : AsyncWrap(env, obj, AsyncWrap::PROVIDER_FILEHANDLE)
+    , buf_(std::move(buf)) {
+  CHECK_NE(source_ = zip_source_buffer_create(buf_.data(), buf_.size(), 0, &error_), nullptr);
+  CHECK_NE(zip_ = zip_open_from_source(source_, 0, &error_), nullptr);
+}
+
+ZipArchive::~ZipArchive() {
+  if (zip_) {
+    CHECK_EQ(zip_close(zip_), 0);
+    zip_ = nullptr;
+  }
+
+  if (source_) {
+    CHECK_EQ(zip_source_close(source_), 0);
+    zip_source_free(source_);
+  }
+}
+
+void ZipArchive::MemoryInfo(MemoryTracker* tracker) const {
+}
+
+void ZipArchive::GetEntries(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+  Environment* env = Environment::GetCurrent(isolate);
+  v8::Local<v8::Context> ctx = env->context();
+
+  ZipArchive* zip;
+  ZipArchive::CheckZip(args, zip);
+
+  const int argc = args.Length();
+  CHECK_EQ(argc, 1);
+
+  CHECK(args[0]->IsBoolean());
+  bool with_file_types = args[0]->IsTrue();
+
+  zip_int64_t num_entries = zip_get_num_entries(zip->zip_, 0);
+  v8::Local<v8::Array> entries = v8::Array::New(isolate, num_entries);
+
+  for (zip_int64_t t = 0; t < num_entries; ++t) {
+    char const* entry_name = zip_get_name(zip->zip_, t, 0);
+    CHECK_NOT_NULL(entry_name);
+
+    v8::Local<v8::Array> entry = v8::Array::New(isolate, 2);
+    entry->Set(ctx, v8::Integer::New(isolate, 0), v8::String::NewFromUtf8(isolate, entry_name).ToLocalChecked());
+    entry->Set(ctx, v8::Integer::New(isolate, 1), v8::Integer::New(isolate, with_file_types ? zip->GetFileType(t) : 0));
+
+    entries->Set(ctx, v8::Integer::New(isolate, t), entry);
+  }
+
+  args.GetReturnValue().Set(entries);
+}
+
+void ZipArchive::AddEntry(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+
+  ZipArchive* zip;
+  ZipArchive::CheckZip(args, zip);
+
+  const int argc = args.Length();
+  CHECK_EQ(argc, 5);
+
+  CHECK(args[0]->IsString());
+  node::Utf8Value path(isolate, args[0]);
+
+  CHECK(args[1]->IsInt32());
+  zip_int32_t comp_method = args[1].As<v8::Int32>()->Value();
+
+  CHECK(Buffer::HasInstance(args[2]));
+  v8::Local<v8::Object> buffer_obj = args[2].As<v8::Object>();
+  char* buffer_data = Buffer::Data(buffer_obj);
+  size_t buffer_length = Buffer::Length(buffer_obj);
+
+  void* buffer_copy = malloc(buffer_length);
```

Comment:
You'd want to use Node.js's allocation functions, i.e. `Malloc()`, `Calloc()` instead of plain `malloc()` (in particular, you don't handle `nullptr` return values here)